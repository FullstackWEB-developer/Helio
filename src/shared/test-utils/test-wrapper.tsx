import configureStore from 'redux-mock-store';
import initialAppState from '@shared/store/app/app.initial-state';
import initialAppUserState from '@shared/store/app-user/appuser.initial-state';
import dayjs from 'dayjs';
import React from 'react';
import {Provider} from 'react-redux';
import {QueryClient, QueryClientProvider} from 'react-query';
import {BrowserRouter} from 'react-router-dom';
import thunk from 'redux-thunk'

export interface TestWrapperProps {
    mockState?: any,
    children: React.ReactNode
}

const TestWrapper = ({ mockState, children} : TestWrapperProps) => {
    const middlewares = [thunk]
    const mockStore = configureStore(middlewares);
    if (!mockState) {
        mockState = {
            appState: {
                ...initialAppState,
                displayLoginRequired: true,
                loginRequiredDismissed: false
            },
            appUserState: {
                ...initialAppUserState,
                expiresOn: dayjs().add(29, 'minutes').toDate()
            }
        };
    }

    const queryClient = new QueryClient({
        defaultOptions: {
            queries: {
                refetchOnWindowFocus: false,
                retry: false
            }
        }
    });

    let store = mockStore(mockState);

    return <Provider store={store}>
        <QueryClientProvider client={queryClient}>
            <BrowserRouter>
                {children}
            </BrowserRouter>
        </QueryClientProvider>
    </Provider>
}

export default TestWrapper;
