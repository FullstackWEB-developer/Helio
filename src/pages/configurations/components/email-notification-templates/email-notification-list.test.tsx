import i18n from '../../../../i18nForTests';
import dayjs from 'dayjs';
import duration from 'dayjs/plugin/duration';
import utc from 'dayjs/plugin/utc';
import customParseFormat from 'dayjs/plugin/customParseFormat';
import {unmountComponentAtNode} from 'react-dom';
import {fireEvent, render, waitFor} from '@testing-library/react';
import TestWrapper from '@shared/test-utils/test-wrapper';
import EmailNotificationList from './email-notification-list';
import Api from '@shared/services/api';
import { Router } from 'react-router';
import { act } from 'react-dom/test-utils';
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useParams: jest.fn(),
   }));
describe("EmailNotificationList tests", () => {
    let container: HTMLDivElement | null;
    let mockState = {
        emailState: {
            unreadEmails: 0,
            messageSummaries: []
        },
        appUserState: {
            liveAgentStatuses: [],
            appUserDetails: {
                id: ""
            }
        },
        lookupsState: {

        },
        appState: {
            smsTemplates: [],
            emailTemplates: []
        }
    };

    beforeEach(async () => {
        await i18n.init();
        dayjs.extend(duration);
        dayjs.extend(utc);
        dayjs.extend(customParseFormat);
        container = document.createElement("div");
        document.body.appendChild(container);
    });

    afterEach(() => {
        if (container) {
            unmountComponentAtNode(container);
            container.remove();
            container = null;
        }
    });

    it("renders EmailNotificationList correctly", async () => {
        let fragment: any;
        await act(async () => {
            const {asFragment} = render(<TestWrapper mockState={mockState}>
                <EmailNotificationList/>          
            </TestWrapper>);
            fragment = asFragment;
        });
        expect(fragment()).toMatchSnapshot();
    });

    it("renders EmailNotificationList with id correctly", async () => {
        var response = {
            data: [{
                id: "string1",
                subject: "string",
                templateBody: "string",
                layoutName: "string",
                name: "string",
                description: "string",
                title: "string",
                defaultBody: "string",
                body: "string",
                createdByName: "string",
                createdOn: new Date(),
                modifiedByName: "string",
                modifiedOn: new Date(),
            },{
                id: "string2",
                subject: "string",
                templateBody: "string",
                layoutName: "string",
                name: "string",
                description: "string",
                title: "string",
                defaultBody: "string",
                body: "string",
                createdByName: "string",
                createdOn: new Date(),
                modifiedByName: "string",
                modifiedOn: new Date(),
            }]
        }
        jest.spyOn(Api, 'get').mockResolvedValue(response);
        await act(async () => {
            const {getByTestId} = render(<TestWrapper mockState={mockState}>
                <EmailNotificationList/>          
            </TestWrapper>);
            await waitFor(() => {
                fireEvent.click(getByTestId('edit-string2'));
            });
        });
        
        expect(Api.get).toHaveBeenCalledTimes(1);
    });

    it("renders EmailNotificationList click sort correctly", async () => {
        var response = {
            data: [{
                id: "string1",
                subject: "string",
                templateBody: "string",
                layoutName: "string",
                name: "string",
                description: "string",
                title: "string",
                defaultBody: "string",
                body: "string",
                createdByName: "string",
                createdOn: new Date(),
                modifiedByName: "string",
                modifiedOn: new Date(),
            },{
                id: "string2",
                subject: "string",
                templateBody: "string",
                layoutName: "string",
                name: "string",
                description: "string",
                title: "string",
                defaultBody: "string",
                body: "string",
                createdByName: "string",
                createdOn: new Date(),
                modifiedByName: "string",
                modifiedOn: new Date(),
            }]
        }
        jest.spyOn(Api, 'get').mockResolvedValue(response);
        await act(async () => {
            const {getByTestId} = render(<TestWrapper mockState={mockState}>
                <EmailNotificationList/>          
            </TestWrapper>);
            await waitFor(() => {
                fireEvent.click(getByTestId('configuration.email_template_list.edit'));
                fireEvent.click(getByTestId('configuration.email_template_list.email_template_name'));
                fireEvent.click(getByTestId('configuration.email_template_list.email_subject'));
                fireEvent.click(getByTestId('configuration.email_template_list.email_template_modified_on'));
            });
        });
        
        expect(Api.get).toHaveBeenCalledTimes(1);
    });

})