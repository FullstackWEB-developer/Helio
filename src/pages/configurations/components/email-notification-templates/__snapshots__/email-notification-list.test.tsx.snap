// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailNotificationList tests renders EmailNotificationList correctly 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col flex-1 overflow-auto h-full p-6 pr-4"
  >
    <h6
      class="pb-7"
    >
      configuration.email_template_list.title
    </h6>
    <div
      class="body2"
    >
      configuration.email_template_list.description
    </div>
    <div
      class="flex flex-col space-y-4 items-center justify-center pt-2 px-2"
    >
      <div>
        <div
          class=""
        >
          <svg
            class="spinner icon-medium "
            viewBox="0 0 40 40"
          >
            <circle
              class="path"
              cx="20"
              cy="20"
              fill="none"
              r="16"
              stroke-width="4"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
